features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/metric/slowframes/UISlownessMetricDispatcher.kt

```diff
@@ -0,0 +1,151 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+package com.datadog.android.rum.internal.metric.slowframes
+
+import androidx.annotation.MainThread
+import androidx.annotation.WorkerThread
+import com.datadog.android.api.InternalLogger
+import com.datadog.android.api.InternalLogger.Target
+import com.datadog.android.rum.configuration.SlowFramesConfiguration
+import com.datadog.tools.annotation.NoOpImplementation
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicInteger
+
+@NoOpImplementation
+internal interface UISlownessMetricDispatcher {
+
+    fun onViewCreated(viewId: String)
+
+    fun incSlowFrame(viewId: String)
+
+    fun incIgnoredFrame(viewId: String)
+
+    fun incFreezeFrame(viewId: String)
+
+    fun sendMetric(viewId: String)
+}
+
+internal class DefaultUISlownessMetricDispatcher(
+    private val config: SlowFramesConfiguration,
+    private val internalLogger: InternalLogger,
+    private val samplingRate: Float = DEFAULT_SAMPLE_RATE
+) : UISlownessMetricDispatcher {
+
+    internal data class SlowFramesTelemetry(
+        var slowFramesCount: AtomicInteger = AtomicInteger(0),
+        var ignoredFramesCount: AtomicInteger = AtomicInteger(0),
+        var freezeFramesCount: AtomicInteger = AtomicInteger(0)
+    )
+
+    private val viewTelemetry = ConcurrentHashMap<String, SlowFramesTelemetry>()
+
+    @MainThread
+    override fun onViewCreated(viewId: String) {
+        viewTelemetry.putIfAbsent(viewId, SlowFramesTelemetry())
+    }
```

Comment:
why there is a `MainThread` requirement here, if collection is backed by `ConcurrentHashMap` anyway?

Reply:
To indicate which thread are usually calls this method. Do you think it worth it to change to comment?

Reply:
yes, this is rather an annotation of which thread _must_ call this thread. Here we can have this be called from any thread, because it has no heavy lifting inside anyway.